### ISSUE TITLE
TypeError when using Calibrator

### ISSUE THREAD

**Issue by 842974287** [NVIDIA/TensorRT#1186]
## Description

Hi, I'm trying to set calibrator to builder but got an error saying `TypeError: Object of type 'NoneType' is not an instance of 'buffer'` when calling `build_cuda_engine`. The cause seems to be returning None in `read_calibration_cache()`. My implementation of the calibrator is like this.

```
class Calibrator(trt.IInt8Calibrator):
    def __init__(self, data, algorithm):
        super().__init__()

        self.data = data
        self.algorithm = algorithm
        self.batch_size = self.data[0][0].shape[0]
        self.count = 0

    def get_batch(self, *args, **kwargs):
        if self.count < len(self.data):
            inputs = self.data[self.count]
            self.count += 1
            return [int(i.contiguous().data_ptr()) for i in inputs]
        else:
            return []

    def get_algorithm(self):
        return self.algorithm

    def get_batch_size(self):
        return self.batch_size

    def read_calibration_cache(self, *args, **kwargs):
        return None

    def write_calibration_cache(self, cache, *args, **kwargs):
        pass
```

## Environment

**TensorRT Version**: 7.2.2.3
**NVIDIA GPU**: P100
**NVIDIA Driver Version**: 450.80.02
**CUDA Version**: 11
**Python Version (if applicable)**: 3.86
**PyTorch Version (if applicable)**: 1.81

## Steps To Reproduce

Use the Calibrator class as calibrator for builder and call `build_cuda_engine()`.

**Comment by pranavm-nvidia** [comment]
Can you try explicitly initializing the base class in your constructor? 
```python
def __init__(self, data, algorithm):
    trt.IInt8Calibrator.__init__(self):
    ...
```

See [the example](https://github.com/NVIDIA/TensorRT/blob/master/samples/python/int8_caffe_mnist/calibrator.py#L51) for details.

You may also want to consider using the [calibrator wrapper provided by Polygraphy](https://docs.nvidia.com/deeplearning/tensorrt/polygraphy/docs/backend/trt/calibrator.html). The corresponding example is [here](https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy/examples/api/04_int8_calibration_in_tensorrt). The example uses Polygraphy throughout, but you can use just the calibrator if you prefer (it behaves exactly like a user-implemented TensorRT calibrator).

**Comment by 842974287** [comment]
@pranavm-nvidia Thanks for the prompt reply! 

Replacing `super().__init__()` with `trt.IInt8Calibrator.__init__(self)` still gives the same error. I'm also trying to set dynamic range manually and have a question about it. I saw in some slides mentioned that TensorRT does symmetric quantization for both weight and activation. I'm wondering for convolution weight if TensorRT is doing per channel quantization or per tensor quantization.

**Comment by pranavm-nvidia** [comment]
@842974287 TensorRT does per-channel quantization for weights. 

Regarding the calibration error, `read_calibration_cache` should have worked fine with returning `None`, but maybe you can try returning an empty buffer?
```python
    def read_calibration_cache(self, *args, **kwargs):
        return b""
```

**Comment by 842974287** [comment]
Hey @pranavm-nvidia, thanks for the suggestion. Returning an empty buffer fixed that error. It seems like I'm getting the error because some changes included in pybind11 version 2.6 updates. 

Those changes caused a problem for `write_calibration_cache` as well. I'm seeing an error here https://github.com/NVIDIA/TensorRT/blob/master/python/src/infer/pyInt8.cpp#L85 because on 2.6+ pybind11 it's start checking if `buffer_info.m_view` is `nullptr` in `py::memoryview` constructor https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L1461. 

Since I'm not actually using those two functions so I commented them out in `pyint8.cpp` and it works fine now.

**Comment by ttyio** [comment]
Closing since we get solution according to last comment, thanks
